src/mpid/ch4/shm/xpmem/xpmem_avl.h

```diff
@@ -0,0 +1,558 @@
+            if (mpi_errno != MPI_SUCCESS)
+                MPIR_ERR_POP(mpi_errno);
+
+            int lheight = sub_root->left == NULL ? 0 : sub_root->left->height;
+            int rheight = sub_root->right == NULL ? 0 : sub_root->right->height;
+            if (lheight - rheight > 1) {
+                /* left unbalanced */
+                MPIDI_XPMEM_seg_t *ct_node = sub_root->left;
+                int lct_height = ct_node->left == NULL ? 0 : ct_node->left->height;
+                if (lct_height + 1 == lheight) {
+                    /* Insert into left child of left child */
+                    /* Do right rotation */
+                    mpi_errno = MPIDI_XPMEM_avl_tree_right_rotation(sub_root, ct_node);
+                    if (mpi_errno != MPI_SUCCESS)
+                        MPIR_ERR_POP(mpi_errno);
+                } else {
+                    /* Insert into right child of left child */
+                    /* Do left-right rotation */
+                    mpi_errno = MPIDI_XPMEM_avl_tree_left_right_rotation(sub_root, ct_node);
+                    if (mpi_errno != MPI_SUCCESS)
+                        MPIR_ERR_POP(mpi_errno);
+                }
+            } else if (rheight - lheight > 1) {
+                /* right unbalanced */
+                MPIDI_XPMEM_seg_t *ct_node = sub_root->right;
+                int lct_height = ct_node->left == NULL ? 0 : ct_node->left->height;
+                if (lct_height + 1 == rheight) {
+                    /* Insert into left child of right child */
+                    /* Do right-left rotation */
+                    mpi_errno = MPIDI_XPMEM_avl_tree_right_left_rotation(sub_root, ct_node);
+                    if (mpi_errno != MPI_SUCCESS)
+                        MPIR_ERR_POP(mpi_errno);
+                } else {
+                    /* Insert into right child of right child */
+                    /* Do left rotation */
+                    mpi_errno = MPIDI_XPMEM_avl_tree_left_rotation(sub_root, ct_node);
+                    if (mpi_errno != MPI_SUCCESS)
+                        MPIR_ERR_POP(mpi_errno);
+                }
+            }
+
+            if (!MPIDI_XPMEM_AVL_TREE_STACK_EMPTY(node_stack)) {
+                MPIDI_XPMEM_AVL_TREE_STACK_POP(node_stack, sub_root);
+                goto stack_recovery;
+            } else
+                break;
+
+        } while (1);
+
+        while (tree->root->parent != NULL)
+            tree->root = tree->root->parent;
+    }
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_XPMEM_SEG_AVL_TREE_INSERT);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_seg_avl_tree_search_and_insert_safe
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_seg_avl_tree_search_and_insert_safe(MPIDI_XPMEM_avl_tree_t
+                                                                             * tree, uint64_t low,
```

Comment:
can we rename all public API to `MPIDI_XPMEM_seg_avl_***` and all internal API to `MPIDI_XPMEM_seg_avl_do_***` ?